### PR TITLE
Rebalances electrocution damage to a log scale

### PR DESCRIPTION
**Pull request by Varen**
## About the PR

Electrocution damage now follows a logarithmic curve. This table shows the values for reference:

| Available Supply  | Damage | Comment |
| ------------- | ------------- | ------------- |
| <=29kw| 0hp | A standard APC is 10kw |
| 30kw | 1hp | A PACMAN generator |
| 50kw | 12hp | A SUPERPACMAN generator |
| 150kw | 34hp | A single SMES or a Substation |
| 600kw | 62hp | 4 SMESes in parallel |
| 1MW | 74hp | ~7 SMESes, or 1 Tesla Coil |
| 2MW | 88hp | ~13 SMESes, or 2 Tesla Coils |
| 4MW | 100hp | ~27 SMESes, or 4 Tesla Coils |
| 10MW | 119hp | ~67 SMESes, or 10 Tesla coils |
| 40MW | 147hp | Just for reference. Logs are fun. |
| 565MW | 200hp | Moth on a lamp. |
| 1.21GW | 216hp | Back to 1955 with you. |

## Why / Balance
Electrified grilles and messing with wires without insulated gloves will now cause damage consistently as of #24554. With the current numbers, this may cause people to be fried for thousands of damage by touching the wrong grille, which is a bit much.

For reference, 4MW would previously cause 80-100 shock before this change. With a MW pulse from some generators, you could get thousands of damage. This mainly peters that part out. See #20880 for other values.

Blue is old curve, Red is new curve.
image

## Technical details
Changed maff. Maff good?

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- tweak: Rebalanced electrocution damage to ramp up slower with power levels.

### CODE REVIEW

**Comment by deltanedas** [comment]
cool gamer electric fence you can make at home!!!!

1. dont have a type G outlet
2. get 2 paperclips
3. put 1 paperclip in live
4. put 1 paperclip in neutral
5. touch both ends with your hands
6. funny N joules through your heart

**Comment by IProduceWidgets** [comment]
Yeah, I don't like how low the low end is. Ideally even a lv wire with an APC should shock you a little.

**Comment by deltanedas** [comment]
electric grilles seem like a short circuit (paperclips in outlet) thing rather than a farmers electric fence so it should do more damage than stunning, and vice versa

**Comment by deltanedas** [comment]
there are actual wire fences which look way nicer for electric fences, grilles are the maint shitter's tool, definitely no smart circuitry there

**Comment by IamVelcroboy** [comment]
The way grilles are used, and the current expected behavior, is as security barriers. If they are changed to not function that way, we should def have an alternative.

**Comment by Partmedia** [comment]
Hi there, I think we're getting a bit lost in the weeds here. For gameplay reasons, electrocution should cause damage, and the damage should discourage people from doing what they did to get electrocuted in the first place.

In other words, electrocution exists to make it harder for you to cut wires and hack doors without getting insulated gloves first. This has the dual purpose of making it harder to grief, and also harder to sabotage.

I think the proposed changes, especially with no damage < 29 kW and very low damage values up to 100 kW make the penalty for electrocution too low.

I tried to make sense of the proposed curve by plotting power on the x-axis and the damage for y. The shape looks about the same but the constants don't quite match up with what was reported in the table. I wonder how you came up with those numbers?

image

Regardless to any concerns about realism, I think it's not okay for LV to do no damage in some cases, and cutting an unlucky HV is an "instacrit".

Considering this, I almost feel like switching back to fixed damage based on LV/MV/HV would be better.

**Comment by Varen** [comment]
> I tried to make sense of the proposed curve by plotting power on the x-axis and the damage for y. The shape looks about the same but the constants don't quite match up with what was reported in the table. I wonder how you came up with those numbers?
For the how part: I've used GeoGebra to do the math and tested it in game at each threshold to confirm I didn't miss anything.
On a hunch, but are you sure you're using the **natural logarithm** in that? MathF.Log uses that.
Excuse the mess I didn't clean it up for showing it publicly but uhh graph.
image
image

>I think the proposed changes, especially with no damage < 29 kW and very low damage values up to 100 kW make the penalty for electrocution too low.
Currently, as it is, under 60k you take no damage. Over 4MW you take enough to crit. This keeps the high around the same, and lowers the low a *bit*, but roughly, it still keeps the LV from being able to electrocute you.
Why? Because that's not in scope for this PR, the main intent is to work on the fact that you can get cryo-or-unrevivable amounts of damage by touching a grille when going in space near the singulo area or the likes. 

As for how much damage LV should do? This is an opinion but, Intuitively, LV roughly serves the same purpose as your home's wiring. And if you're working on that, the insulation of the tool handle itself is plenty enough.

I will once again reference you to the PR that had the previous numbers for comparison: https://github.com/space-wizards/space-station-14/pull/20880#issuecomment-1755899624

**Comment by Partmedia** [comment]
> For the how part: I've used GeoGebra to do the math and tested it in game at each threshold to confirm I didn't miss anything. On a hunch, but are you sure you're using the **natural logarithm** in that? MathF.Log uses that. Excuse the mess I didn't clean it up for showing it publicly but uhh graph.

Yes, I'm using the natural logarithm. I'm probably missing some constant scaling factors somewhere.

You can make it really easy to review by plotting the existing curve on top of the proposed curve :)

> > I think the proposed changes, especially with no damage < 29 kW and very low damage values up to 100 kW make the penalty for electrocution too low.
> 
> Currently, as it is, under 60k you take no damage. Over 4MW you take enough to crit. This keeps the high around the same, and lowers the low a _bit_, but roughly, it still keeps the LV from being able to electrocute you. Why? Because that's not in scope for this PR, the main intent is to work on the fact that you can get cryo-or-unrevivable amounts of damage by touching a grille when going in space near the singulo area or the likes.

I wasn't aware of what is currently implemented in the game, and since the proposal is lower than what currently exists in the game, I think it's an improvement.

> As for how much damage LV should do? This is an opinion but, Intuitively, LV roughly serves the same purpose as your home's wiring. And if you're working on that, the insulation of the tool handle itself is plenty enough.

I stated the purpose of the electrocution mechanic. Let me repeat that again:

> In other words, electrocution exists to make it harder for you to cut wires and hack doors without getting insulated gloves first. This has the dual purpose of making it harder to grief, and also harder to sabotage.

I think it should be high enough to achieve those stated goals.

> I will once again reference you to the PR that had the previous numbers for comparison: [#20880 (comment)](https://github.com/space-wizards/space-station-14/pull/20880#issuecomment-1755899624)
> 
> On another one TEST WHY YOU FAIL I DON'T GET D:

I re-ran the tests for you.

Ultimately, this seems to sound better than what currently exists in game and therefore I intend to merge it. If you can provide that plot, that should make a pretty strong case for this change.

File: Content.Server/Electrocution/ElectrocutionSystem.cs [hunk 1]
```diff
@@ -62,9 +62,10 @@ public sealed class ElectrocutionSystem : SharedElectrocutionSystem
     [ValidatePrototypeId<DamageTypePrototype>]
     private const string DamageType = "Shock";
 
-    // Yes, this is absurdly small for a reason.
-    private const float ElectrifiedScalePerWatt = 1E-6f;
-
+    // Multiply and shift the log scale for shock damage.
+    private const float ElectrifiedScaleLogCoefficent = 1.35f;
+    private const float ElectrifiedShiftDamage = 1.67f;
+    private const float ElectrifiedDamageSupplyCoefficent = 0.00001f;
```
[the remark this answers is not in the document]
Comment (Varen) [reply Content.Server/Electrocution/ElectrocutionSystem.cs]: Changed things around so it's all in the same function, and added some explanation for each part of it.
Also added an image on the PR to show old VS new damage model.

**Comment by Partmedia** [comment]
Thank you for your contribution!

**Pull request merged**